### DISCUSSION TITLE
Problem with aspect ratio

### DISCUSSION THREAD

**Discussion by asmaaharfoush** [neperfepx/neper#205] [Q&A] [answered]
Hello, I tried to use different values of  aspect ratio in creating 2D microstructures (1,1), (1,4) and (1,7), the problem is that the figure of grain for (1,1) looks too elongated the elongation decrease when I (1,4)

Command for (1,1)
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01),aspratio(1,1)' -o 0 -id 0 -dim 2 -domain 'square(1,1):rotate(0,0,1,25):scale(0.5,2,1)' -transform 'scale(2,0.5,1),rotate(0,0,1,-25 )'
neper -V 0.tess -cameraangle 12 -imagesize 256:256 -dataedgerad 0.001 -print 0
![0](https://user-images.githubusercontent.com/37200158/116878169-d088af80-abd3-11eb-9b36-07782836bb43.png)

Command for (1,4)
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01),aspratio(1,4)' -o 3 -id 3 -dim 2 -domain 'square(1,1):rotate(0,0,1,25):scale(0.5,2,1)' -transform 'scale(2,0.5,1),rotate(0,0,1,-25 )'
neper -V 3.tess -cameraangle 12 -imagesize 256:256 -dataedgerad 0.001 -print 3
![3](https://user-images.githubusercontent.com/37200158/116878467-455be980-abd4-11eb-83cb-95c903663c76.png)

Command for (1,7)
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01),aspratio(1,7)' -o 6 -id 6 -dim 2 -domain 'square(1,1):rotate(0,0,1,25):scale(0.5,2,1)' -transform 'scale(2,0.5,1),rotate(0,0,1,-25 )'
neper -V 6.tess -cameraangle 12 -imagesize 256:256 -dataedgerad 0.001 -print 6
![6](https://user-images.githubusercontent.com/37200158/116878623-7a683c00-abd4-11eb-9434-f7643972b054.png)

I am not what is the reason that makes the figure does not match the aspect ratio required, do you have any insight?
Thanks!

**Comment by rquey** [comment, marked as answer]
This discussion follows [this comment](https://github.com/rquey/neper/discussions/185#discussioncomment-674741). In the command I provided, note that I _did not_ use `aspratio` to generate elongated grains. This is because it is not appropriate when the elongation direction is other than the sample directions (x, y, z). Instead, I used a combination of scalings (to get the aspect ratio) and rotations (to get them in the required directions). See [this discussion](https://github.com/rquey/neper/discussions/196) for details on `aspratio`.

In your case, do not use `aspratio` and play with the parameters of the `scale` transformations:
```
for i in 1 4 7
do
  val1=`echo "sqrt($i)" | bc -l`
  val2=`echo "1/sqrt($i)" | bc -l`

  neper -T -n 1000 -dim 2 -domain "square(1,1):rotate(0,0,1,25):scale($val2,$val1,1)" -transform "scale($val1,$val2,1),rotate(0,0,1,-25)" -o 1-$i
  neper --rcfile none -V 1-$i.tess -dataedgerad 0.002 -datacellcol id -imagesize 600:600 -cameraangle 12 -print 1-$i
done
```
![1-1](https://user-images.githubusercontent.com/2659376/117313645-5f245780-ae86-11eb-9e1c-020c5da09bc5.png)
![1-4](https://user-images.githubusercontent.com/2659376/117313640-5e8bc100-ae86-11eb-84b8-d093c62f2629.png)
![1-7](https://user-images.githubusercontent.com/2659376/117313637-5df32a80-ae86-11eb-8aac-369decabdeff.png)

**Reply by asmaaharfoush** [reply]
Thanks for your reply, Is that could be reachable if I use lognoraml distribution for morpho?

**Reply by rquey** [reply]
Yes, of course.

**Reply by asmaaharfoush** [reply]
Thank you for your help!

- I just wanted to double-check because I got confused about some concepts, this part of the code -domain "square(1,1):rotate(0,0,1,25):scale($val2,$val1,1)" work on rotating and elongating the domain (square),  regarding to the second part of the code -transform "scale($val1,$val2,1),rotate(0,0,1,-25)" , you are trying put the domain in a regular square is that correct? **We are making change mainly on the domain what is exactly the relationship between the grain inside and the change that we did?**

- Can the values (val2, val1) be considered as aspect ratio? If I want an aspect ratio of 1:3 can I  put val1 =1, and val2 =3

- Why have you disabled  looking at the initialization file by using this command --rcfile none

**Reply by rquey** [reply]
> Thank you for your help!
>
> I just wanted to double-check because I got confused about some concepts, this part of the code -domain "square(1,1):rotate(0,0,1,25):scale($val2,$val1,1)" work on rotating and elongating the domain (square),  regarding to the second part of the code -transform "scale($val1,$val2,1),rotate(0,0,1,-25)" , you are trying put the domain in a regular square is that correct? **We are making change mainly on the domain what is exactly the relationship between the grain inside and the change that we did?**

See https://github.com/rquey/neper/discussions/196.

>    Can the values (val2, val1) be considered as aspect ratio? If I want an aspect ratio of 1:3 can I  put val1 =1, and val2 =3

No, because there is the 3rd direction. Make the factors so that they multiply to 1. If you want a ratio of 3 between the 1st and 2nd values, use sqrt(3) and 1/sqrt(3)...

>    Why have you disabled  looking at the initialization file by using this command --rcfile none

Not to mess up the visualization with my own initialization file.

**Reply by asmaaharfoush** [reply]
Thank you for your reply, you mean that we need to use sqrt(3) and 1/sqrt(3), using the code you provided above and we can get the aspect ratio 1:3, is that correct?

**Reply by rquey** [reply]
The grains will be scaled by sqrt(3) in one direction and 1/sqrt(3) in the other, which does give a 1:3 ratio.

**Comment by asmaaharfoush** [comment]
Hello,

Thanks for helping me, I wrote a small code that changes the morpho, rotation, and ratio. It does working pretty well  except for some values such this command that does not generate a figure:

```
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01)' -o 4 -id 4 -dim 2 -domain 'square(1,1):rotate(0,0,1,45):scale(0.5773502691896258,1.7320508075688772,1)' -transform 'scale(1.7320508075688772,0.5773502691896258,1),rotate(0,0,1,-45 )'
neper -V 4.tess -imagesize 256:256 -dataedgerad 0.001 -print 4
```

I don't have any explanation why the previous command does not work however the former and latter commands working properly: 
```
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01)' -o 3 -id 3 -dim 2 -domain 'square(1,1):rotate(0,0,1,25):scale(0.5773502691896258,1.7320508075688772,1)' -transform 'scale(1.7320508075688772,0.5773502691896258,1),rotate(0,0,1,-25 )'
neper -V 3.tess -imagesize 256:256 -dataedgerad 0.001 -print 3
```

```
neper -T -n from_morpho -morpho 'diameq:normal(0.07,0.01)' -o 5 -id 5 -dim 2 -domain 'square(1,1):rotate(0,0,1,65):scale(0.5773502691896258,1.7320508075688772,1)' -transform 'scale(1.7320508075688772,0.5773502691896258,1),rotate(0,0,1,-65 )'
neper -V 5.tess -imagesize 256:256 -dataedgerad 0.001 -print 5
```

Do you have any explanation for that, thank you!
